activerecord/lib/active_record/insert_all.rb

```diff
@@ -14,7 +14,7 @@ def initialize(model, inserts, on_duplicate:, returning: nil, unique_by: nil)
       @returning = (connection.supports_insert_returning? ? primary_keys : false) if @returning.nil?
       @returning = false if @returning == []
 
-      @unique_by = find_unique_index_for(unique_by) if unique_by
+      @unique_by = find_unique_index_for(unique_by || model.primary_key)
```

Comment:
It's my understanding that it's not always we want to apply a `unique_by` clause.

Reply:
If the `unique_by` is matching the primary_key - `find_unique_index_for` returns nil - so the update only occurs if the `on_duplicate == :update`.

The reason that is it getting passed in to get the same behaviour for passing in a primary key as you get if you to an `upsert` out specifying the `unique_by`.

This might be better expressed as:

```
unique_by ||= model.primary_key if @on_duplicate == :update
```

And having find_unique_index_for returning a representation of the primary key if the primary key matches.